File: python/src/tdcv2/engine/router.py
"""Which engine runs a config.

A config does not name an engine — it states a CONSTRAINT, and the router picks the fastest engine
that can honour it. ``mode="memory"`` means the whole run may be held at once; ``mode="disk"``
means it may not, and the choice between the two streaming engines follows from what the config
actually asks for. Naming an engine outright with ``engine="1|2|3"`` is available and skips all of
this, which is what makes it useful for a benchmark and a poor default for everything else.

The interesting decisions are the ones that route a disk-mode config BACK to memory. Each marks
something whose answer depends on the whole column — an interpolated pack address, an exact share
declared inside a pack, a weighted draw of a linked row. Answered a row at a time they do not fail;
they quietly produce data that is wrong in a way nobody notices, which is the worst outcome
available and the reason these checks exist.
"""

from __future__ import annotations

from collections.abc import Callable

from ..generators import advanced_regex
from ..model.config import Config, Gen
from ..packs import DataPacks


def resolve(config: Config, packs: DataPacks | None = None) -> int:
    """The engine a config runs on: 1 in memory, 2 streaming, 3 exact on disk."""
    forced = _trim_to_none(config.engine)
    mode_for_conflict = _trim_to_none(config.mode)
    # `engine=` wins over `mode=` — except when the two contradict. `mode="sequential"`
    # is not a preference about speed, it is a promise that row N is computed after row
    # N-1, which only engine 1 keeps. Letting `engine="2"` override it silently produced
    # the worst possible message: a run failing with "add mode=sequential" to a config
    # that already said it.
    if mode_for_conflict == "sequential" and forced is not None and forced != "1":
        raise ValueError(
            f'engine="{forced}" contradicts mode="sequential": rows must be computed in '
            "order, and only engine 1 does that. Drop one of the two."
        )
    if forced is not None:
        if forced not in ("1", "2", "3"):
            raise ValueError(
                f'invalid engine "{forced}" — expected "1" (in-memory), "2" (streaming), '
                'or "3" (exact-on-disk)'
            )
        return int(forced)

    mode = _trim_to_none(config.mode)
    if mode == "memory":
        return 1
    # Rows strictly in order, so `prev()` has a previous row. Engine 2 resolves ANY row
    # in O(1) without touching the one before it — that is its design, and it is what
    # this mode cannot be built on. The cost is engine 1's: the run is held in memory.
    if mode == "sequential":
        return 1
    if mode == "stream":
        # The old name for asking for Engine 2 outright, from before mode described the constraint
        # rather than the engine. Kept working; the router is not consulted.
        return 2
    if mode is not None and mode != "disk":
        raise ValueError(f'invalid mode "{mode}" — expected "memory", "disk" or "sequential"')
    # No mode at all means disk: a config says how big its run is, not how to hold it, and the
    # engine that can stream is the right default for a generator whose whole point is volume.

    # A template address that names a field resolves per row against the other columns; only the
    # in-memory engine has them all.
    if _any_gen(config, lambda gen: gen.type == "template" and _is_dynamic(gen.attr("value"))):
        return 1
    # weight= with row= draws a linked record to an exact quota, which needs the global total.
    if _any_gen(
        config,
        lambda gen: (
            gen.type == "file"
            and _trim_to_none(gen.attrs.get("weight")) is not None
            and _trim_to_none(gen.attrs.get("row")) is not None
        ),
    ):
        return 1
    # A pack generator that declares its own shares apportions them over the whole column.
    if packs is not None and _any_gen(config, lambda gen: _declares_shares(gen, config, packs)):
        return 1
    # uniq on a DRAWN value takes WITHOUT REPLACEMENT — simple or composed alike — the pool and the
    # taken-set span the whole column, which only the in-memory engine holds.
    if any(
        s.uniq
        and (
            s.gen.type not in ("increment", "decrement")
            if s.gen is not None
            else any(
                i.gen is not None
                and i.field is None
                and i.gen.type not in ("increment", "decrement")
                for i in (s.items or [])
            )
        )
        for s in config.sequences
    ):
        return 1
    # A network call is not reproducible, so it never runs on the reproducible path.
    if _any_gen(config, lambda gen: gen.type == "http"):
        return 1
    # A <switch> branch that declares a share the streaming engines cannot lay over the right
    # rows. They refuse such a branch rather than apportion it over the wrong denominator, and a
    # refusal reached at build time is not a fallback for every caller. Decide it here,
    # statically, where every path sees the same answer.
    if _unstreamable_switch_percent(config):
        return 1
    return 3 if needs_exact(config) else 2


def _case_carries_percent(case) -> bool:
    """Does this ``<case>`` body declare a share that the denominator has to be right for?"""
    if case is None:
        return False
    return any(
        (part.mix is not None and (part.mix.percent or "").strip() != "")
        or (part.gen is not None and part.gen.attr("percent").strip() != "")
        for part in case.parts
    )


def _unstreamable_switch_percent(config: Config) -> bool:
    """A ``<switch>`` branch whose share the streaming engines cannot honour.

    They can subset a branch keyed on ONE value of a plain values list — the same bijection
    ``parent="Gender.Male"`` uses. They cannot rank a multi-key branch (``US|CA|MX`` is a union,
    and ranks across a union do not compose from the per-value ranks), nor ``<default>`` (a
    complement, which nothing enumerates), nor any branch whose subject is not a finite values
    list to begin with.

    Deliberately conservative: anything it cannot prove streamable goes to engine 1, which costs
    speed on an exotic config and never costs correctness.
    """
    by_name = {spec.name: spec for spec in config.sequences}

    def plain_list_values(name: str) -> list[str] | None:
        subject = by_name.get(name)
        gen = None if subject is None else subject.gen
        if gen is None or gen.type != "text":
            return None
        if gen.attr("order") == "sequential" or gen.attr("repeat").strip() != "":
            return None
        return [v.strip() for v in gen.attr("value").split(",")]

    def nested_switches(case) -> list:
        """Every ``<switch>`` written inside this ``<case>`` body, at any depth."""
        found: list = []

        def visit(body) -> None:
            if body is None:
                return
            for part in body.parts:
                if part.switch is not None:
                    found.append(part.switch)
                    for entry in part.switch.entries:
                        visit(entry.value)
                    visit(part.switch.fallback)
                elif part.mix is not None:
                    for inner in part.mix.cases:
                        visit(inner)

        visit(case)
        return found

    for spec in config.sequences:
        sw = spec.switch_spec
        bodies = []
        if sw is not None:
            bodies.extend(entry.value for entry in sw.entries)
            bodies.append(sw.fallback)
        if spec.mix is not None:
            bodies.extend(spec.mix.cases)
        # A NESTED switch is never rankable — its branch covers an intersection of two
        # partitions, and there is no O(1) rank inside one. So any share it declares, at any
        # depth, decides engine 1.
        for body in bodies:
            for nested in nested_switches(body):
                if _case_carries_percent(nested.fallback) or any(
                    _case_carries_percent(e.value) for e in nested.entries
                ):
                    return True

        if sw is None:
            continue
        if _case_carries_percent(sw.fallback):
            return True
        values = plain_list_values(sw.on)
        for entry in sw.entries:
            if not _case_carries_percent(entry.value):
                continue
            if len(entry.keys) != 1 or values is None or entry.keys[0] not in values:
                return True
    return False


def needs_exact(config: Config) -> bool:
    """Whether disk mode needs the exact engine rather than the streaming one.

    Everything here is a case where a per-row answer and a whole-column answer differ: ANY
    uniqueness, a child of a parent whose values are not a finite list, a weighted choice inside
    a pattern. Ordinary exact percentages, switch, distinct and text parent-child all stream.

    ``uniq`` is here in full, and that is a deliberate cost. A group REARRANGES the columns it
    covers — every column keeps its multiset, so every declared share survives — and that cannot
    be decided a row at a time. The streaming engine could only offer a different answer, and two
    answers from one seed is the thing this whole design exists to prevent.
    """
    by_name = {spec.name: spec for spec in config.sequences}

    if config.env_uniq_groups:
        return True

    for spec in config.sequences:
        if spec.uniq:
            return True
        if spec.gen is not None and _is_weighted_advanced_regex(spec.gen):
            return True
        for f in _fields_of(spec):
            if _is_weighted_advanced_regex(f.gen):
                return True
        parent = _trim_to_none(spec.parent)
        if parent is not None and not _parent_is_finite_text(by_name, parent):
            return True
    return False


def _parent_is_finite_text(by_name, reference: str) -> bool:
    dot = reference.find(".")
    parent = by_name.get(reference if dot < 0 else reference[:dot])
    return parent is not None and parent.gen is not None and parent.gen.type == "text"


def _is_weighted_advanced_regex(gen: Gen) -> bool:
    return gen.type == "advanced_regex" and advanced_regex.has_weighted_choice(gen.attr("value"))


def _has_percent(gen: Gen) -> bool:
    return bool(gen.attrs.get("percent"))


def _declares_shares(gen: Gen, config: Config, packs: DataPacks) -> bool:
    if gen.type != "template":
        return False
    path = gen.attr("value")
    if not path or _is_dynamic(path):
        return False
    locale = gen.attrs.get("local") or config.locale
    try:
        return _needs_whole_column(packs, path, locale)
    except (ValueError, OSError):
        # An address that does not resolve is the validator's problem, not the router's.
        return False


def _needs_whole_column(packs: DataPacks, path: str, locale: str) -> bool:
    """Whether a pack GENERATOR apportions a share over the whole column.

    Two ways to earn it.

    A ``percent=`` anywhere in a generator's body — on its ``<mix>``, on a ``<gen>``, on a compound
    field — makes its quota a property of the run rather than of a row.

    And a body that DRAWS from a weighted list, which the body does not say. A weighted list is
    laid out to an exact Hamilton quota over the run, so each value takes its measured share of the
    rows; asked for one row, that plan is computed over a column of one and the row goes to the
    largest share, every time, for every seed. Twelve shipped full-name packs were in that state —
    ``hu.person.male.fullName`` returned the same name on all forty rows of a forty-row run —
    because the body says ``value="hu.person.lastName"`` and nothing in that line says the list on
    the other end carries weights.

    Only a generator is ever marked. A weighted list drawn DIRECTLY by a config is not this
    problem: the streaming engine lays one out across the column already, and marking it here
    would take `en.person.lastName` off the streaming engines for nothing. The weighted leaf
    matters one level down, inside a body, which is what ``_draws_whole_column`` is for.

    The router asks this before sending a config to an engine that resolves one row at a time.
    """
    entry = packs.load(path, locale)
    if not entry.is_generator:
        return False
    return _body_needs_whole_column(packs, entry.generator or "", locale, frozenset({path}))


def _body_needs_whole_column(
    packs: DataPacks, body: str, locale: str, seen: frozenset[str]
) -> bool:
    """The share test over one generator body: its own ``percent=``, or what it draws from."""
    from ..parser import config_builder

    try:
        composed = config_builder.parse_pack_body(body)
    except (ValueError, AttributeError):
        # A single bare <gen>: it declares a share only through its own percent=.
        return "percent=" in body

    for spec in composed.sequences:
        if spec.is_mix and spec.mix.percent and spec.mix.percent.strip():
            return True
        for gen in [spec.gen, *(f.gen for f in _fields_of(spec))]:
            if gen is None:
                continue
            if _has_percent(gen):
                return True
            if _draws_whole_column(packs, gen, locale, seen):
                return True
    return False


def _draws_whole_column(packs: DataPacks, gen, locale: str, seen: frozenset[str]) -> bool:
    """Does a ``<gen>`` in a pack body reach a weighted list, directly or through another pack?"""
    if gen.type != "template":
        return False
    target = gen.attrs.get("value", "")
    if not target or _is_dynamic(target) or target in seen:
        return False  # `target in seen` is the cycle guard; the pack loader reports the cycle
    inner = gen.attrs.get("local") or locale
    try:
        if not packs.exists(target, inner):
            return False
        entry = packs.load(target, inner)
        # HERE a weighted list is the leaf the walk is looking for — it is the thing the body
        # draws from, and its quota is what the body inherits.
        if entry.weighted:
            return True
        if not entry.is_generator:
            return False
        return _body_needs_whole_column(packs, entry.generator or "", inner, seen | {target})
    except (ValueError, OSError):
        return False


def _is_dynamic(value: str) -> bool:
    """``common.vehicle.model.${{Brand}}`` — an address that is not known until the row is."""
    return "${{" in value


def _any_gen(config: Config, test: Callable[[Gen], bool]) -> bool:
    """Every ``<gen>`` in the config, simple or a compound's field."""
    for spec in config.sequences:
        if spec.gen is not None and test(spec.gen):
            return True
        for f in _fields_of(spec):
            if f.gen is not None and test(f.gen):
                return True
    return False


def _fields_of(spec):
    """A compound's fields, or nothing — a simple sequence has none rather than an empty list."""
    return spec.fields if spec.is_compound and spec.fields else []


def _trim_to_none(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None

File: rust/src/engine/router.rs
//! Which engine a config gets.
//!
//! The three engines agree: one config, one seed, one set of bytes, whichever
//! engine renders it — that is what `fixtures/cross-language/engines.json` pins
//! down. What differs is what each engine can do at all, and at what cost in
//! memory. So routing is still part of the contract, but for a narrower reason:
//! a config sent to an engine that cannot answer its question does not fail, it
//! answers a smaller question — an exact share becomes a per-row guess — and
//! the output is wrong in every row while looking perfectly plausible.
//!
//! A config does not name an engine — it states a constraint, and the router
//! picks the fastest engine that can honour it. `mode="memory"` means the whole
//! run may be held at once; `mode="disk"` means it may not. Naming an engine
//! outright with `engine="1|2|3"` skips all of this, which is what makes it
//! useful for a benchmark and a poor default for everything else.
//!
//! The interesting decisions are the ones that route a disk-mode config back to
//! memory. Each marks something whose answer depends on the whole column — an
//! interpolated pack address, an exact share declared inside a pack, a weighted
//! draw of a linked row. Answered a row at a time they do not fail; they quietly
//! produce data that is wrong in a way nobody notices, which is the worst
//! outcome available and the reason these checks exist.

use super::{invalid, EngineResult};
use crate::generators::advanced_regex;
use crate::model::{Config, Gen, SequenceSpec, Source, Switch};
use crate::packs::DataPacks;

/// The engine a config runs on: 1 in memory, 2 streaming, 3 exact on disk.
///
/// `packs` is what the pack-driven rule reads — a generator declares its shares
/// inside its own file, so there is no way to see them without opening it.
/// `None` skips that one rule and answers from the config alone, which is what a
/// caller that has no registry to hand can honestly do.
pub fn resolve(config: &Config, packs: Option<&DataPacks>) -> EngineResult<u8> {
    // `engine=` wins over `mode=` — except when the two contradict. `mode="sequential"`
    // is not a preference about speed, it is a promise that row N is computed after row
    // N-1, which only engine 1 keeps. Letting `engine="2"` override it silently produced
    // the worst possible message: a run failing with "add mode=sequential" to a config
    // that already said it.
    if trim_to_none(config.mode.as_deref()) == Some("sequential") {
        if let Some(forced) = trim_to_none(config.engine.as_deref()) {
            if forced != "1" {
                return invalid(&format!(
                    "engine=\"{forced}\" contradicts mode=\"sequential\": rows must be \
                     computed in order, and only engine 1 does that. Drop one of the two."
                ));
            }
        }
    }
    if let Some(forced) = trim_to_none(config.engine.as_deref()) {
        return match forced {
            "1" => Ok(1),
            "2" => Ok(2),
            "3" => Ok(3),
            other => invalid(&format!(
                "invalid engine \"{other}\" — expected \"1\" (in-memory), \"2\" (streaming), \
                 or \"3\" (exact-on-disk)"
            )),
        };
    }

    match trim_to_none(config.mode.as_deref()) {
        Some("memory") => return Ok(1),
        // Rows strictly in order, so `prev()` has a previous row. Engine 2 resolves ANY
        // row in O(1) without touching the one before it — that is its design, and it is
        // what this mode cannot be built on. The cost is engine 1's: the run is held in
        // memory.
        Some("sequential") => return Ok(1),
        // "stream" is the old name for asking for Engine 2 outright, from before
        // mode described the constraint rather than the engine. Kept working;
        // the router is not consulted.
        Some("stream") => return Ok(2),
        Some(other) if other != "disk" => {
            return invalid(&format!(
                "invalid mode \"{other}\" — expected \"memory\", \"disk\" or \"sequential\""
            ))
        }
        // No mode at all means disk: a config says how big its run is, not how
        // to hold it, and the engine that can stream is the right default for a
        // generator whose whole point is volume.
        _ => {}
    }

    // A template address that names a field resolves per row against the other
    // columns; only the in-memory engine has them all.
    if any_gen(config, |gen| {
        gen.gen_type == "template" && is_dynamic(gen.attr_or("value", ""))
    }) {
        return Ok(1);
    }

    // weight= with row= draws a linked record to an exact quota, which needs the
    // global total.
    if any_gen(config, |gen| {
        gen.gen_type == "file"
            && trim_to_none(gen.attr("weight")).is_some()
            && trim_to_none(gen.attr("row")).is_some()
    }) {
        return Ok(1);
    }

    // A pack generator that declares its own shares used to be routed here, and for
    // a real reason: resolved a row at a time the quota was computed over a single
    // row and every row went to the largest share. The streaming builder plans such
    // a body over the COLUMN now, so the reason is gone — and keeping the rule after
    // the refusal went was worse than nothing, because the config still landed on
    // the engine that holds the whole table. Measured on a 5,000,000-row
    // `hu.person.male.fullName` column: the in-memory engine wanted 2 GB and died
    // under a 512 MB cap, while the streaming path finished it inside 512 MB.
    //
    // One shape still belongs here — a body carrying its own `<valid>` — and it
    // arrives the way every other unstreamable config does: refused by name.

    // A network call is not reproducible, so it never runs on the reproducible
    // path.
    // uniq on a DRAWN value takes WITHOUT REPLACEMENT — simple or composed alike — the pool and the
    // taken-set span the whole column, which only the in-memory engine holds.
    let counting = |t: &str| t == "increment" || t == "decrement";
    if config.sequences.iter().any(|s| {
        use crate::model::config::{Item, Source};
        s.uniq
            && match &s.source {
                Source::Gen(gen) => !counting(&gen.gen_type),
                Source::Items(items) => items
                    .iter()
                    .any(|i| matches!(i, Item::Gen(gen) if !counting(&gen.gen_type))),
                _ => false,
            }
    }) {
        return Ok(1);
    }
    if any_gen(config, |gen| gen.gen_type == "http") {
        return Ok(1);
    }
    // A <switch> branch that declares a share the streaming engines cannot lay over the right
    // rows. They refuse such a branch rather than apportion it over the wrong denominator, and a
    // refusal reached at build time is not a fallback for every caller. Decide it here,
    // statically, where every path sees the same answer.
    if unstreamable_switch_percent(config) {
        return Ok(1);
    }

    Ok(if needs_exact(config) { 3 } else { 2 })
}

/// Does this `<case>` body declare a share that the denominator has to be right for?
fn case_carries_percent(case: Option<&crate::model::Case>) -> bool {
    case.is_some_and(|c| {
        c.parts.iter().any(|part| match part {
            crate::model::CasePart::Mix(mix) => {
                !mix.percent.as_deref().unwrap_or("").trim().is_empty()
            }
            crate::model::CasePart::Gen(gen) => !gen.attr_or("percent", "").trim().is_empty(),
            // A nested switch declares no share of its own; each of ITS branches is judged
            // separately, in `unstreamable_switch_percent`.
            crate::model::CasePart::Text(_) | crate::model::CasePart::Switch(_) => false,
        })
    })
}

/// A `<switch>` branch whose share the streaming engines cannot honour.
///
/// They can subset a branch keyed on ONE value of a plain values list — the same bijection
/// `parent="Gender.Male"` uses. They cannot rank a multi-key branch (`US|CA|MX` is a union, and
/// ranks across a union do not compose from the per-value ranks), nor `<default>` (a complement,
/// which nothing enumerates), nor any branch whose subject is not a finite values list.
///
/// Deliberately conservative: anything it cannot prove streamable goes to engine 1, which costs
/// speed on an exotic config and never costs correctness.
/// Every `<switch>` written inside this `<case>` body, at any depth.
fn nested_switches(case: &crate::model::Case, found: &mut Vec<Switch>) {
    for part in &case.parts {
        match part {
            crate::model::CasePart::Switch(sw) => {
                found.push((**sw).clone());
                for entry in &sw.entries {
                    nested_switches(&entry.value, found);
                }
                if let Some(fallback) = &sw.fallback {
                    nested_switches(fallback, found);
                }
            }
            crate::model::CasePart::Mix(mix) => {
                for inner in &mix.cases {
                    nested_switches(inner, found);
                }
            }
            _ => {}
        }
    }
}

fn unstreamable_switch_percent(config: &Config) -> bool {
    let plain_list_values = |name: &str| -> Option<Vec<String>> {
        let subject = config.sequences.iter().find(|s| s.name == name)?;
        let gen = match &subject.source {
            Source::Gen(gen) => gen,
            _ => return None,
        };
        if gen.gen_type != "text"
            || gen.attr_or("order", "") == "sequential"
            || !gen.attr_or("repeat", "").trim().is_empty()
        {
            return None;
        }
        Some(
            gen.attr_or("value", "")
                .split(',')
                .map(|v| v.trim().to_string())
                .collect(),
        )
    };

    // A NESTED switch is never rankable — its branch covers an intersection of two partitions,
    // and there is no O(1) rank inside one. So any share it declares, at any depth, decides
    // engine 1.
    let nested_declares_share = |bodies: Vec<&crate::model::Case>| -> bool {
        let mut found = Vec::new();
        for body in bodies {
            nested_switches(body, &mut found);
        }
        found.iter().any(|nested| {
            case_carries_percent(nested.fallback.as_ref())
                || nested
                    .entries
                    .iter()
                    .any(|e| case_carries_percent(Some(&e.value)))
        })
    };

    if config.sequences.iter().any(|spec| match &spec.source {
        Source::Switch(sw) => {
            let mut bodies: Vec<&crate::model::Case> =
                sw.entries.iter().map(|e| &e.value).collect();
            if let Some(fallback) = &sw.fallback {
                bodies.push(fallback);
            }
            nested_declares_share(bodies)
        }
        Source::Mix(mix) => nested_declares_share(mix.cases.iter().collect()),
        _ => false,
    }) {
        return true;
    }

    config.sequences.iter().any(|spec| {
        let Source::Switch(sw) = &spec.source else {
            return false;
        };
        if case_carries_percent(sw.fallback.as_ref()) {
            return true;
        }
        let values = plain_list_values(&sw.on);
        sw.entries.iter().any(|entry| {
            case_carries_percent(Some(&entry.value))
                && (entry.keys.len() != 1
                    || values
                        .as_ref()
                        .is_none_or(|vs| !vs.contains(&entry.keys[0])))
        })
    })
}

/// Whether disk mode needs the exact engine rather than the streaming one.
///
/// Everything here is a case where a per-row answer and a whole-column answer
/// differ: exact percentages combined with uniqueness, a uniq field that is not
/// a finite list, a child of a parent whose values are not a finite list, a
/// weighted choice inside a pattern. Ordinary exact percentages, uniform
/// uniqueness, switch, distinct and text parent-child all stream.
pub fn needs_exact(config: &Config) -> bool {
    // A group REARRANGES the columns it covers — every column keeps its multiset, so every
    // declared share survives — and that cannot be decided a row at a time. The streaming
    // engine could only offer a different answer, and two answers from one seed is the thing
    // this whole design exists to prevent.
    if !config.env_uniq_groups.is_empty() {
        return true;
    }

    for spec in &config.sequences {
        if spec.uniq {
            return true;
        }

        if let Source::Gen(gen) = &spec.source {
            if is_weighted_advanced_regex(gen) {
                return true;
            }
        }
        if fields_of(spec).into_iter().any(is_weighted_advanced_regex) {
            return true;
        }

        if let Some(parent) = trim_to_none(spec.parent.as_deref()) {
            if !parent_is_finite_text(config, parent) {
                return true;
            }
        }
    }
    false
}

fn parent_is_finite_text(config: &Config, reference: &str) -> bool {
    let name = match reference.find('.') {
        Some(dot) => &reference[..dot],
        None => reference,
    };
    config
        .sequence(name)
        .and_then(SequenceSpec::gen)
        .is_some_and(|gen| gen.gen_type == "text")
}

/// A `<gen type="advanced_regex">` whose pattern weights its branches —
/// `(?%{70:RU;20:US;10:DE})`.
///
/// A weighted choice is an exact share over the whole column, so it cannot be
/// answered one row at a time. Decided by parsing the pattern and counting, not
/// by searching for the opener: `(?%{` can also appear inside a character class,
/// where it is four ordinary characters.
fn is_weighted_advanced_regex(gen: &Gen) -> bool {
    gen.gen_type == "advanced_regex"
        && advanced_regex::has_weighted_choice(gen.attr_or("value", ""))
}

/// A `<gen type="template">` naming a pack generator that declares a share.
///
/// A dynamic address is left to the rule above it — the pack it names is not
/// known here, and that config is already on its way to engine 1.

/// `common.vehicle.model.${{Brand}}` — an address not known until the row is.
fn is_dynamic(value: &str) -> bool {
    value.contains("${{")
}

/// Every `<gen>` in the config, simple or a compound's field.
fn any_gen(config: &Config, test: impl Fn(&Gen) -> bool) -> bool {
    config.sequences.iter().any(|spec| match &spec.source {
        Source::Gen(gen) => test(gen),
        Source::Fields(fields) => fields.iter().any(|f| test(&f.gen)),
        _ => false,
    })
}

/// A compound's fields, or nothing — a simple sequence has none rather than an
/// empty list.
fn fields_of(spec: &SequenceSpec) -> Vec<&Gen> {
    match &spec.source {
        Source::Fields(fields) => fields.iter().map(|f| &f.gen).collect(),
        _ => Vec::new(),
    }
}

fn trim_to_none(value: Option<&str>) -> Option<&str> {
    let trimmed = value?.trim();
    (!trimmed.is_empty()).then_some(trimmed)
}
